refactor(indexer): optimize indexes on optimistic tables

kodemartin opened the pull request:
# Description of change

This patch does the following changes:

* Extends the `tx_global_order` table to include the `tx_sequence_number` assigned to transactions when checkpointed. This is to allow for more efficient pagination queries, as we observed 20-fold decrease in performance.
* It also adds the composite key `(global_sequence_number, optimistic_sequence_number)` to `optimistic_{tx,event}_indices` to allow for more efficient joins with `tx_global_order`

## Links to any relevant issues

Part of #5871 

## How the change has been tested


- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [x] Synchronization of the indexer from genesis for a network including migration objects.
- [x] Restart of indexer synchronization locally without resetting the database.
- [x] Restart of indexer synchronization on a production-like database.
- [x] Deployment of services using Docker.
- [x] Verification of API backward compatibility.

kodemartin committed:
refactor(indexer): extend tx_global_order with chk sequence number

kodemartin committed:
refactor(indexer): update keys and constraings for {tx,event}_indices

tomxey commented on crates/iota-indexer/migrations/pg/2025-05-29-072412_insertion-order-revisited/up.sql:
```diff
@@ -19,25 +19,152 @@ DROP TABLE IF EXISTS tx_insertion_order;
+ALTER TABLE optimistic_event_struct_instantiation  ADD PRIMARY KEY (package, module, type_instantiation, global_sequence_number, optimistic_sequence_number, event_sequence_number);
+ALTER TABLE optimistic_event_senders               ADD PRIMARY KEY (sender, global_sequence_number, optimistic_sequence_number, event_sequence_number);
+
+-- optimistic_event_emit_package
```
foreign keys for optimistic_tx_* seem to be missing

tomxey commented on crates/iota-indexer/migrations/pg/2025-05-29-072412_insertion-order-revisited/up.sql:
```diff
@@ -19,25 +19,152 @@ DROP TABLE IF EXISTS tx_insertion_order;
+ALTER TABLE optimistic_event_struct_instantiation  DROP CONSTRAINT optimistic_event_struct_instantiation_tx_insertion_order_fkey;
+ALTER TABLE optimistic_event_senders               DROP CONSTRAINT optimistic_event_senders_tx_insertion_order_fkey;
+
+ALTER TABLE optimistic_transactions                DROP CONSTRAINT optimistic_transactions_pkey;
```
apart from primary and foreign keys, there were also some indexes defined on the tx_* and event_* tables that I believe also should be updated to include both optimistic and global number. (e.g. optimistic_tx_changed_objects_sender)

tomxey commented on crates/iota-indexer/src/models/events.rs:
```diff
@@ -106,7 +109,7 @@ impl From<IndexedEvent> for StoredEvent {
     fn from(event: OptimisticEvent) -> Self {
         Self {
-            tx_sequence_number: event.sequence_number,
+            tx_sequence_number: event.global_sequence_number,
```
It's a bit asymmetric that here we populate `tx_sequence_number` from `global_seq_num`, but in the `from_stored` function we populate `optimistic_sequence_number` from `tx_sequence_number`.

It probably doesn't matter though.

tomxey commented on crates/iota-indexer/src/models/transactions.rs:
```diff
@@ -40,6 +40,7 @@ use crate::{
 #[derive(Clone, Debug, Queryable, Insertable, QueryableByName, Selectable)]
 #[diesel(table_name = tx_global_order)]
 pub struct TxGlobalOrder {
+    pub chk_tx_sequence_number: Option<i64>,
```
a short comment may be useful, since we have 3 sequence numbers on the same model

tomxey commented on crates/iota-indexer/src/models/transactions.rs:
```diff
@@ -182,7 +186,8 @@ impl From<OptimisticTransaction> for StoredTransaction {
 impl From<StoredTransaction> for OptimisticTransaction {
```
Should we remove this, since we have the `from_stored` function implemented?

kodemartin committed:
refactor(indexer): remove doc ref to private struct

tomxey committed:
Drop and recreate optimistic tables

tomxey committed:
Minor review remarks:
- remove unused From implementation
- use optimistic_sequence_number when converting from Optimistic to Stored models
- add doc to chk_tx_sequence_number field

sergiupopescu199 reviewed (approved):
lgtm!
